File: android/app/src/main/java/com/airbridge/app/transfer/TransferQueue.kt
package com.airbridge.app.transfer

import com.airbridge.app.core.interfaces.ITransferSession
import com.airbridge.app.core.interfaces.TransferState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Holds and processes [ITransferSession] objects, running up to [concurrency]
 * sessions in parallel.
 *
 * Sessions are dispatched in enqueue order within each concurrency slot.
 * [pauseAll] pauses every active session.
 * [cancelAll] cancels every session that has not yet reached a terminal state.
 *
 * @param concurrency Maximum number of sessions that may run simultaneously. Defaults to 1.
 * @param scope       Optional [CoroutineScope] to run sessions in. Defaults to a private
 *                    IO-dispatcher scope that is cancelled when [close] is called.
 */
class TransferQueue(
    val concurrency: Int = 1,
    scope: CoroutineScope? = null,
) : AutoCloseable {

    init {
        require(concurrency >= 1) { "concurrency must be >= 1, was $concurrency" }
    }

    private val _scope = scope ?: CoroutineScope(Dispatchers.IO + SupervisorJob())

    /** Bounded channel used as the work queue (one slot per pending item). */
    private val _channel = Channel<ITransferSession>(capacity = Channel.UNLIMITED)

    private val _mutex    = Mutex()
    private val _sessions = mutableListOf<ITransferSession>()

    /** Snapshot of all sessions that have been enqueued (regardless of state). */
    val allSessions: List<ITransferSession> get() = _sessions.toList()

    init {
        // Launch [concurrency] worker coroutines that drain the channel.
        repeat(concurrency) {
            _scope.launch {
                for (session in _channel) {
                    runSession(session)
                }
            }
        }
    }

    /**
     * Adds [session] to the queue.  Returns the [Job] for the enqueue operation
     * (not the session itself — state tracking is via [ITransferSession.stateFlow]).
     */
    suspend fun enqueue(session: ITransferSession): Job {
        _mutex.withLock { _sessions.add(session) }
        _channel.send(session)
        return _scope.coroutineContext[Job]!!
    }

    /**
     * Pauses all sessions that are currently [TransferState.ACTIVE].
     */
    suspend fun pauseAll() {
        val snapshot = _mutex.withLock { _sessions.toList() }
        for (s in snapshot) {
            if (s.stateFlow.replayOrNull() == TransferState.ACTIVE) s.pause()
        }
    }

    /**
     * Cancels every session that has not yet reached a terminal state.
     */
    suspend fun cancelAll() {
        val snapshot = _mutex.withLock { _sessions.toList() }
        for (s in snapshot) {
            when (s.stateFlow.replayOrNull()) {
                TransferState.PENDING,
                TransferState.ACTIVE,
                TransferState.PAUSED -> s.cancel()
                else -> Unit
            }
        }
    }

    // ── AutoCloseable ──────────────────────────────────────────────────────

    override fun close() {
        _channel.close()
        _scope.cancel()
    }

    // ── Private ────────────────────────────────────────────────────────────

    private suspend fun runSession(session: ITransferSession) {
        try {
            session.start()
        } catch (_: Exception) {
            // Session state is already set to FAILED or CANCELLED.
            // Errors are observable via stateFlow — do not propagate here so that
            // the worker keeps processing subsequent sessions.
        }
    }
}

// ── Extension helpers ──────────────────────────────────────────────────────────

/**
 * Returns the most-recently emitted value of this [kotlinx.coroutines.flow.StateFlow],
 * or null if this is not a StateFlow (graceful fallback for test mocks).
 */
private fun <T> kotlinx.coroutines.flow.Flow<T>.replayOrNull(): T? =
    (this as? kotlinx.coroutines.flow.StateFlow<T>)?.value

File: windows/AirBridge.Transfer/TransferQueue.cs
using AirBridge.Core.Interfaces;

namespace AirBridge.Transfer;

/// <summary>
/// Holds and processes <see cref="ITransferSession"/> objects, running up to
/// <see cref="Concurrency"/> sessions in parallel.
/// <para>
/// Sessions are started in enqueue order within each concurrency slot.
/// <see cref="PauseAllAsync"/> pauses all active sessions; <see cref="CancelAllAsync"/>
/// cancels everything, including queued sessions that have not yet started.
/// </para>
/// </summary>
public sealed class TransferQueue : IDisposable
{
    private readonly int _concurrency;
    private readonly SemaphoreSlim _slot;
    private readonly List<ITransferSession> _allSessions = new();
    private readonly object _lock = new();
    private bool _disposed;

    /// <summary>
    /// Initializes a new <see cref="TransferQueue"/>.
    /// </summary>
    /// <param name="concurrency">
    /// Maximum number of sessions that may run concurrently. Defaults to 1 (sequential).
    /// </param>
    public TransferQueue(int concurrency = 1)
    {
        if (concurrency < 1) throw new ArgumentOutOfRangeException(nameof(concurrency));
        _concurrency = concurrency;
        _slot        = new SemaphoreSlim(concurrency, concurrency);
    }

    /// <summary>Maximum number of sessions that run simultaneously.</summary>
    public int Concurrency => _concurrency;

    /// <summary>
    /// Snapshot of all sessions that have been enqueued (regardless of state).
    /// </summary>
    public IReadOnlyList<ITransferSession> AllSessions
    {
        get { lock (_lock) return _allSessions.ToList(); }
    }

    /// <summary>
    /// Adds <paramref name="session"/> to the queue and schedules it to run when a
    /// concurrency slot is available.  Returns a <see cref="Task"/> that completes
    /// when the session reaches a terminal state (<c>Completed</c>, <c>Failed</c>,
    /// or <c>Cancelled</c>).
    /// </summary>
    /// <param name="session">The session to enqueue.</param>
    /// <param name="cancellationToken">
    /// Token that cancels the session if signalled while it is waiting for a slot
    /// or while it is running.
    /// </param>
    public Task EnqueueAsync(ITransferSession session, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(session);
        lock (_lock) _allSessions.Add(session);
        return RunWhenSlotAvailableAsync(session, cancellationToken);
    }

    /// <summary>
    /// Pauses all sessions that are currently in the <c>Active</c> state.
    /// Already-paused, completed, or cancelled sessions are not affected.
    /// </summary>
    public async Task PauseAllAsync()
    {
        List<ITransferSession> snapshot;
        lock (_lock) snapshot = _allSessions.ToList();

        foreach (var s in snapshot)
            if (s.State == TransferState.Active)
                await s.PauseAsync().ConfigureAwait(false);
    }

    /// <summary>
    /// Cancels every session that has not yet reached a terminal state.
    /// </summary>
    public async Task CancelAllAsync()
    {
        List<ITransferSession> snapshot;
        lock (_lock) snapshot = _allSessions.ToList();

        foreach (var s in snapshot)
            if (s.State is TransferState.Pending or TransferState.Active or TransferState.Paused)
                await s.CancelAsync().ConfigureAwait(false);
    }

    // ── Private ────────────────────────────────────────────────────────────

    private async Task RunWhenSlotAvailableAsync(ITransferSession session, CancellationToken ct)
    {
        // Wait for a free concurrency slot (or cancellation)
        await _slot.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            if (ct.IsCancellationRequested)
            {
                await session.CancelAsync().ConfigureAwait(false);
                return;
            }

            // Let the session-level CancellationToken propagate; if it throws we
            // swallow it here so the queue keeps processing subsequent sessions.
            try
            {
                await session.StartAsync(ct).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                // Session was cancelled or paused — state is already set by the session.
            }
            catch (Exception ex)
            {
                // Log-friendly: surface as a completed (faulted) task; the session State
                // is already Failed. Callers who care attach StateChanged handlers.
                _ = ex; // suppress unused-variable warning
            }
        }
        finally
        {
            _slot.Release();
        }
    }

    // ── IDisposable ────────────────────────────────────────────────────────

    /// <inheritdoc/>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _slot.Dispose();
    }
}
